packages/grafana-ui/src/components/SelectNG/SelectMenu.tsx

```diff
@@ -0,0 +1,156 @@
+export const SelectMenu = React.forwardRef<HTMLDivElement, SelectMenuProps>((props, ref) => {
+  const {
+    maxHeight = 300,
+    options,
+    highlightedIndex,
+    selectedItem,
+    getItemProps,
+    enableOptionCreation,
+    onOptionCreate,
+    inputValue,
+    renderOption,
+    ...otherProps
+  } = props;
+  const styles = useStyles(getSelectStyles);
+
+  let groupLengthAggr = 0;
+  let flatOptionsLengthAggr = 0;
+  // Maps options so SelectMenuOption components
+  const optionElements = options.map((o, index) => {
+    const itemPropsGetter = (options: GetItemPropsOptions<any>) =>
+      getItemProps({
+        'aria-label': selectors.components.Select.option,
+        key: o.value,
+        isSelected: o === selectedItem,
+        ...options,
+        //  disabled: TODO
+      });
+
+    const optionEl = renderOption(o, itemPropsGetter, index + groupLengthAggr, highlightedIndex);
+
+    if (o.options) {
+      groupLengthAggr += o.options.length - 1; // -1 to compensate indexing from 0
+    } else {
+      flatOptionsLengthAggr += 1;
+    }
+
+    return optionEl;
+  });
+
+  // When creating custom options is enabled append link to do so
+  if (enableOptionCreation && shouldAllowOptionCreate(options, inputValue)) {
+    const newItem = { value: inputValue, label: inputValue };
+    console.log(flatOptionsLengthAggr);
+    const itemProps = getItemProps({
+      key: inputValue,
+      item: newItem,
+      index: groupLengthAggr + flatOptionsLengthAggr,
+    });
+
+    optionElements.push(
+      <SelectMenuOption
+        {...itemProps}
+        key="custom-options"
+        onClick={(e: React.MouseEvent) => {
+          e.stopPropagation();
+          if (onOptionCreate) {
+            onOptionCreate(newItem);
+          }
+          itemProps.onClick(e);
+        }}
+        isFocused={highlightedIndex === options.length}
+        item={{ value: inputValue, label: `Create: ${inputValue}` }}
+      />
+    );
+  }
+
+  return (
+    <div
+      className={styles.menu}
+      style={{ maxHeight }}
+      {...otherProps}
+      ref={ref}
+      aria-label={selectors.components.Select.menu}
+    >
+      <CustomScrollbar autoHide={false} autoHeightMax="inherit" hideHorizontalTrack>
+        {optionElements}
+      </CustomScrollbar>
+    </div>
+  );
+});
+
+SelectMenu.displayName = 'SelectMenu';
+
+interface SelectMenuOptionProps<T> extends React.HTMLAttributes<HTMLDivElement> {
+  disabled?: boolean;
+  isFocused?: boolean;
+  isSelected?: boolean;
+  renderOptionLabel?: (value: SelectableValue<T>) => JSX.Element;
+  item: SelectableValue<T>;
+}
+
+export const SelectMenuOption: React.FC<SelectMenuOptionProps<any>> = props => {
+  const theme = useTheme();
+  const styles = getSelectStyles(theme);
+  const { children, item, renderOptionLabel, isSelected, isFocused, ...otherProps } = props;
+  return (
+    <div {...otherProps} className={cx(styles.option, isFocused && styles.optionFocused)}>
```

Comment:
Not relevant here but I think this'd be a perfect use case for using `styled` as it'd abstract all this css login into a separate component.